### PR TITLE
[process-agent] Add loopback resolution

### PR DESCRIPTION
**Pull request by hmahmood**
### What does this PR do?

Add resolution for loopback to loopback flows

### Motivation

This is required for visibility within pods

### Describe your test plan
1. Setup istio on a k8s cluster
2. Deploy an nginx pod with istio injected
3. curl the pod IP of the nginx pod
4. In flow explorer, check for loopback to loopback connections, dest port 80, source port ephemeral. These should resolve to source container as the istio sidecar proxy container and the dest container as the nginx container.

### CODE REVIEW

File: pkg/process/checks/net.go [hunk 1]
```diff
@@ -230,7 +327,7 @@ func connectionPIDs(conns []*model.Connection) []int32 {
 }
 
 // getCtrIDsByPIDs will fetch container id and pid relationship from either process check or container check, depend on which one is enabled and ran
```
Comment (heyronhay) [inline pkg/process/checks/net.go:329]: Need to update the function name in the comment

**Review by heyronhay — APPROVED** [review]
looks good to these newbie eyes - just one comment fix

File: pkg/process/checks/net.go [hunk 2]
```diff
@@ -36,6 +36,11 @@ type ConnectionsCheck struct {
 	lastTelemetry          *model.CollectorConnectionsTelemetry
 }
 
+type addrWithNS struct {
+	model.ContainerAddr
+	ns uint32
```
Comment (sunhay) [inline pkg/process/checks/net.go:41]: `ns` -> `netns` for consistency where we use this elsewhere?

File: pkg/process/checks/net.go [hunk 3]
```diff
@@ -85,12 +90,102 @@ func (c *ConnectionsCheck) Run(cfg *config.AgentConfig, groupID int32) ([]model.
...
+
+	resolveLoopbackConnections(conns.Conns, ctrsByPid)
+
 	tel := c.diffTelemetry(conns.Telemetry)
 
 	log.Debugf("collected connections in %s", time.Since(start))
 	return batchConnections(cfg, groupID, c.enrichConnections(conns.Conns), conns.Dns, c.networkID, tel), nil
 }
 
+func resolveLoopbackConnections(conns []*model.Connection, ctrsByPid map[int32]string) {
```
Comment (sunhay) [inline pkg/process/checks/net.go:104]: Thoughts on moving this logic somewhere within `pkg/network` and calling out to it from here? This network check is getting a little unwieldy

Comment (sunhay) [reply pkg/process/checks/net.go:104]: I noticed we used `LocalResolver.LoadAddrs(ctrList)` below - sounds like this might be a good fit to live within the `LocalResolver`?

File: pkg/process/checks/net.go [hunk 4]
```diff
@@ -85,12 +90,102 @@ func (c *ConnectionsCheck) Run(cfg *config.AgentConfig, groupID int32) ([]model.
...
+
+	ctrsByLaddr := make(map[model.ContainerAddr]string)
+	ctrsByLoopback := make(map[addrWithNS]string)
```
Comment (leeavital) [inline pkg/process/checks/net.go:107]: It might be clearner to combine this into a single (pre-sized) map and use 0 as the netns for n on-loopback addrs  ?

**Review by leeavital — COMMENTED** [review]
Don't forget to fill in the "test plan" portion of the PR description, so we can test this before releasing.

File: pkg/process/checks/net.go [hunk 5]
```diff
@@ -85,12 +90,102 @@ func (c *ConnectionsCheck) Run(cfg *config.AgentConfig, groupID int32) ([]model.
...
+			ctrsByLaddr[claddr] = cid
+		}
+	}
+
+	log.Tracef("ctrsByLoopback = %v", ctrsByLoopback)
+	log.Tracef("ctrsByLaddr = %v", ctrsByLaddr)
+
+	for _, conn := range conns {
+		if conn.Raddr.ContainerId != "" {
+			log.Tracef("skipping already resolved raddr %v", conn.Raddr)
```
Comment (leeavital) [inline pkg/process/checks/net.go:140]: Is all this logging useful for development, or should we keep it in the final version?

File: pkg/process/checks/net.go (hunk 3 above)
Comment (hmahmood) [reply pkg/process/checks/net.go]: Yes, good idea, done.

File: pkg/process/checks/net.go (hunk 4 above)
Comment (hmahmood) [reply pkg/process/checks/net.go]: Yes, done.

File: pkg/process/net/resolver/resolver_test.go [hunk 6]
```diff
@@ -1,9 +1,11 @@
 package resolver
 
 import (
+	"fmt"
 	"net"
 	"testing"
 
+	"github.com/DataDog/agent-payload/process"
```
Comment (brycekahle) [inline pkg/process/net/resolver/resolver_test.go:8]: nit: This import is already available as `model`

File: pkg/process/net/resolver/resolver_test.go [hunk 7]
```diff
@@ -97,3 +98,233 @@ func TestLocalResolver(t *testing.T) {
...
+		"10.1.1.1:1234":  "foo1",
+		"127.0.0.1:1235": "foo3",
+		"1.2.3.4:1234":   "bar",
+		"127.0.0.1:1250": "foo7",
+		"127.0.0.1:1260": "foo6",
+	}
+
+	resolver.Resolve(&model.Connections{Conns: tests})
+
+	for _, te := range tests {
```
Comment (p-lambert) [inline pkg/process/net/resolver/resolver_test.go:305]: I appreciate the amount of effort in writing these tests and making the assertion messages very clear, but I found it super hard to follow.

Perhaps you could have something like
```go
type testCase struct {
	c *model.Connection
	expectedLaddrID string
	expectedRaddrID string
}
```

and structure each test case like:
```go
testCase{
	c: *model.Connection{
		{
			Pid: 1,
			Laddr: &model.Addr{
				Ip:   "127.0.0.1",
				Port: 1234,
			},
			Raddr: &model.Addr{
				Ip:   "10.1.1.2",
				Port: 1234,
			},
			IpTranslation: &model.IPTranslation{
				ReplDstIP:   "10.1.1.1",
				ReplDstPort: 1234,
				ReplSrcIP:   "10.1.1.2",
				ReplSrcPort: 1234,
			},
			NetNS: 1,
		},
	},
	expectedLaddrID: "foo",
	expectedRaddrID: "bar",
}
```
to make it more readable?

**Review by p-lambert — APPROVED** [review]
Looks great to me. I left one nit about the tests (feel free to ignore it though).
Apart from that, could you perhaps update the comments on
```go
func (l *LocalResolver) Resolve(c *model.Connections)
```
And provide some context/explanation about what it is being done there?

File: pkg/process/net/resolver/resolver_test.go (hunk 7 above)
Comment (hmahmood) [reply pkg/process/net/resolver/resolver_test.go:305]: Yes, that is a good suggestion. Let me make the changes.

**Comment by hmahmood** [comment]
> Looks great to me. I left one nit about the tests (feel free to ignore it though).
> Apart from that, could you perhaps update the comments on
> 
> ```go
> func (l *LocalResolver) Resolve(c *model.Connections)
> ```
> 
> And provide some context/explanation about what it is being done there?

Done, let me what you think @p-lambert

**Review by leeavital — APPROVED** [review]
